README.adoc
image:https://travis-ci.org/jorgefranconunes/tea.svg?branch=master["Build Status", link="https://travis-ci.org/jorgefranconunes/tea"]
image:https://codecov.io/gh/jorgefranconunes/tea/branch/master/graph/badge.svg[codecov, link="https://codecov.io/gh/jorgefranconunes/tea"]





== The Tea Scripting Language

Tea is a scripting language for the Java environment.

This document contains instructions on how to setup the development
environment for working on Tea itself. It is intended for the Tea
development team.

This development tree is for the 4.y.z master development branch.





== Pre-Requisites

* JDK 1.8
* Maven 3.3+





== Preparing the Development Environment

These instructions apply to unix-like environments. If you really want to
develop under windows you should have Cygwin (http://www.cygwin.com/)
installed on your system.


Fetch the source tree from the Git repository hosted at GitHub

----
git clone git@github.com:jorgefranconunes/tea.git
----





== Building

To perform a full build just run `mvn` in the `modules` dir of the
working area.

By default unit tests are not run. To run the unit tests as part of
the build define the system property `skipTests=false`, like this:

----
cd ./modules
mvn -DskipTests=false
----


In addition there are some Maven profiles that can be activated to
perform specific tasks.

* `-P tea-tunit-tests` : Runs TUnit tests, even when `skipTests=true`
   (which is, by default).

* `-P tea-teadoc` : Creates Teadoc documentation from the project
   source files. The Teadoc documentation is created under
   'doc/teadoc'. This is tipically used when building a release.

* `-P tea-javadoc` : Creates Javadoc documentation from the project
   source files. The Javadoc documentation is created under
   'doc/javadoc'. This is tipically used when building a release.





== Making a New Release

. Check the release version number. Ensure that the version number is
coherent in `config/tea-core.conf` and `pom.xml`.

. Build the release bundle by invoking Maven like this:
+
----
mvn -Dtea.release
----

. The release tarbal in the project root directory as
'./target/tea-4.x.y.tar.bz2'.





== Bug reports

Send your bug reports and requests for new features to
tea.dev@pdmfc.com



doc/history.txt
2011/03/31 (jpsl) TSK-PDMFC-TEA-0056 The build process now only
strictly requires ant 1.7. Build instructions included for the eclipse
and netbeans Java IDEs.




2010/10/04 (jpsl) TSK-PDMFC-TEA-0034 teaEngine-0.7.0 merged into
Tea 4.0.0beta2 trunk. JSR-223 API is now part of the standard Tea
release. Please read doc/engineProgGuide.html for details.




2008/08/30 (jfn) Simplified the configuration of the build
environment.

See "00README-developers.txt" for details on the new simplified
procedure for seting up the development environment.





2008/08/28 (jfn) Created branch 3.2.x on the Subversion repository.

The 3.2.x branch is intended for maintenance only. No new features
should be added in this branch. Only bug fixes.

svn mkdir -m 'Created repository directory for branches.' \
	svn+ssh://www.pdmfc.com/opt/develop/svnroot/tea/branches

svn cp -m 'Created branch 3.2.x with development tree as of 2007/06/04.' \
        -r 1 \
	svn+ssh://www.pdmfc.com/opt/develop/svnroot/tea/trunk \
	svn+ssh://www.pdmfc.com/opt/develop/svnroot/tea/branches/3.2.x


The 3.2.x branch was created in revision 17.

To create a local copy of the 3.2.x branch:

cd ~/wherever
cvn svn+ssh://www.pdmfc.com/opt/develop/svnroot/tee/branches/3.2.x tea

The above commands will create inside "~/wherever" a directory named
"tea" with the latest contents of the 3.2.x branch.





2001/04/05

In a Tea program run by the "tsh" utility a global variable named
"argv0" is now created right before executing the Tea script
file. This variable contains the path to the script file that is being
executed.





2001/02/17

Tea 2 has been officialy launched. The Tea home page will soon be
updated with the news and a link to the download page.

CVS tag: TEA020000





2001/02/15

Started development on Tea version 2. A new CVS repository was setup
having CVSROOT=:pserver:tdk@red.solsuni.pt:/opt/develop/cvsroot





2001/02/08

Added the "file-make-path" function. (jfn)





2001/01/19

Added the "tea-autoload-package" function.

Tea functions and classes of the TDBC module are now loaded on demand.





2001/01/10

The TeaDoc DTD has the following new elements:

Class
ClassRef
MethodRef
FuncRef

The "teadoc" tool was updated to support these new elements.





2000/12/15

The "and" function now may accept no arguments, as mentioned in the
documentation.

Added the "file-unlink-recursive" function. It removes a directory and
all its contents.





2000/12/10

Added support for variables shared among all the Tea interpreters
running in a JVM. New functions in the "Lang" module:

tea-shared-define
tea-shared-defined?
tea-shared-get
tea-shated-set!


Added support for synchronization between different Tea interpreters
running in different threads. New functions in the "Lang" module:

tea-lock-acquire
tea-lock-release


The Tea site (http://www.devweb.pt/tea) uses these features to
maintain track of the current number of visitors to the site.






2000/11/27

The "TInput copyTo" method can now receive an aditional argument
containing the maximum number of bytes to read from the input stream.

The "TSocket copyTo" method can also receive that aditional argument.


Released the Tea Development Kit version 1.2.2. Major differences from
the previous release:

    <ul>
    <li>The <code>TInput</code> and <code>TSocket</code> classes
    <code>copyTo</code> methods now accept an aditional optional byte
    count argument.</li>
    <li>The utilities provided with the TDK (tsh, teadoc) have been
    moved to a single bin directory (<code>apps/bin</code>).</li>
    </ul>





2000/11/18

bug-0012 (copyTo pending flush) solved.





2000/11/17

Released the Tea Development Kit version 1.2.1. Major differences from
the previous release:

    <ul>
    <li>Solved vicious bug that prevented the correct flush of
    <code>TOutput</code> streams in some particular
    circumstances. This also affected <code>TSocket</code>.</li>
    <li>New reflection functions in the TOS module
    (<code>class-of</code>, <code>class-base-of</code>,
    <code>class-is-a</code>, <code>class-get-name</code>).</li>
    <li>Runtime error messages envolving TOS classes are now much more
    informative.</li>
    </ul>






2000/11/15

All classes of the Tea language now have a name. Error messages
envolving classes are now much more clear.

Added new functions in the TOS module:

class-of
class-base-of
class-is-a
class-get-name





2000/11/11

Released the Tea Development Kit 1.2.0.  Major differences from
the previous release:

    <ul>
    <li>Added support for server side sockets
    (<code>TServerSocket</code>).</li>
    <li>The Tea parser now supports the same escape sequences as the
    Java language.</li>
    <li>New <code>str-unescape</code> function for transforming a
    string containing Tea escape sequences.</li>
    </ul>





2000/11/09

Tea error messages in a Tea program about unknown methods now refer
the name of the TOS class invoved.

Consider the following piece of code:

class x () ; [new x] abc

When running this code the error message would have been:

class 'STOSClass' has no method 'abc'

Now it is the following:

class 'x' has no method 'abc'





2000/11/08

Added the "str-unescape" function ("String" module).

Added the class "TServerSocket" ("Net" module) for using TCP/IP server
sockets.





2000/11/05

The following escape sequences are now supported by the Tea compiler:

       \b backspace
       \f form feed
       \n newline
       \r carriage return
       \t horizontal tab
       \" double quote
       \' single quote
       \\ backslash
       \xxx character, in octal (000-377)
       \uxxxx Unicode character, in hexadecimal (0000-FFFF)





2000/10/05

Added the "file-size" function ("IO" module).





2000/10/03

Added support for methods with a variable number of arguments.

Added runtime support to byte arrays.

The "TInput copyTo" now only requires the object received as argument

src/com/pdmfc/tea/compiler/SCode.java
/**************************************************************************
 *
 * Copyright (c) 2001 PDM&FC, All Rights Reserved.
 *
 **************************************************************************/

/**************************************************************************
 *
 * $Id: SCode.java,v 1.7 2004/04/02 19:39:06 jfn Exp $
 *
 *
 * Revisions:
 *
 * 2002/03/28
 * The SCode instances now store the file name from where they
 * were compiled. (jfn)
 *
 * 2001/05/12
 * Created. (jfn)
 *
 **************************************************************************/

package com.pdmfc.tea.compiler;

import java.io.PrintStream;
import java.util.Enumeration;

import com.pdmfc.tea.STeaException;
import com.pdmfc.tea.compiler.SStatement;
import com.pdmfc.tea.runtime.SContext;
import com.pdmfc.tea.runtime.SFlowControlException;
import com.pdmfc.tea.runtime.SObjNull;
import com.pdmfc.tea.runtime.SRuntimeException;





/**************************************************************************
 *
 * A Tea program that can be executed in a given context. A Tea
 * program is just a sequence of Tea statements.
 *
 **************************************************************************/

public class SCode
    extends Object {




    SCodeNode _statsHead = null;
    SCodeNode _statsTail = null;
    int       _statCount = 0;
    String    _fileName  = null;

    private static final String ERR_STAT =
	"While evaluating statement on line {0}";

    private static final String ERR_STAT_FILE =
	"While evaluating statement on line {0} ({1})";





/**************************************************************************
 *
 * Builds an empty statement list.
 *
 **************************************************************************/

    SCode(String fileName) {

	_fileName = fileName;
    }





/**************************************************************************
 *
 * Executes the sequence of Tea statements in the <TT>context</TT>
 * given as argument.
 *
 * @param context Context inside wich the block will be executed.
 *
 * @return The value returned by the last command in the block or null
 * if it is an empty block.
 *
 * @exception STeaException Only thrown by the execution of one of the
 * statements.
 *
 **************************************************************************/

    public final Object exec(SContext context)
	throws STeaException {

	Object value = SObjNull.NULL;

	for ( SCodeNode node=_statsHead; node!=null; node=node._next ) {
	    SStatement statement = node._element;
	    try {
		value = statement.exec(context);
	    } catch (SRuntimeException e) {
		int      lineNum = statement.getLineNumber();
		Object[] fmtArgs = { String.valueOf(lineNum), _fileName };
		String   fmtMsg  = (_fileName==null) ? ERR_STAT :ERR_STAT_FILE;

		e.addMessage(fmtMsg, fmtArgs);
		throw e;
	    } catch (SFlowControlException e) {
		throw e;
	    } catch (STeaException e) {
		int      lineNum = statement.getLineNumber();
		Object[] fmtArgs = { String.valueOf(lineNum), _fileName };
		String   fmtMsg  = (_fileName==null) ? ERR_STAT :ERR_STAT_FILE;
		SRuntimeException error =new SRuntimeException(e.getMessage());

		error.addMessage(fmtMsg, fmtArgs);
		throw error;		
	    }
	}

	return value;
    }






/**************************************************************************
 *
 * Adds a new Tea statement at the end of the sequence.
 *
 * @param aStatement Reference to the new statement to be appended to
 * the sequence.
 *
 **************************************************************************/

    final void addStatement(SStatement aStatement) {

	SCodeNode newStat = new SCodeNode(aStatement);

	if ( _statsHead == null ) {
	    _statsHead = newStat;
	} else {
	    _statsTail._next = newStat;
	}
	_statsTail = newStat;
	_statCount++;
    }






/**************************************************************************
 *
 * This is used for debugging. At the moment it is rather slow.
 *
 * @param out A stream where the message will be sent to.
 *
 * @param ident Number of padding white spaces inserted into the left.
 *
 **************************************************************************/

//    public final void prettyPrint(PrintStream out, int indent) {
//
//	for (  SCodeNode node=_statsHead; node!=null; node=node._next ) {
//	    node._element.prettyPrint(out, indent);
//	}
//    }


}





/**************************************************************************
 *
 * Nodes of a linked list.
 *
 **************************************************************************/

class SCodeNode
    extends Object {





    public SStatement _element = null;
    public SCodeNode  _next    = null;





/**************************************************************************
 *
 * 
 *
 **************************************************************************/

    public SCodeNode(SStatement statement) {

	_element = statement;
    }

}





/**************************************************************************
 *
 * 
 *
 **************************************************************************/

